### DISCUSSION TITLE
How does Include and Exclude really work in sync_list?

### DISCUSSION THREAD

**Discussion by Flakvard** [abraunegg/onedrive#2369] [Q&A] [answered]
### First of all this is a fantastic app!
I really like it and made it much easier for me to switch to linux from windows.

### Understanding Include and Exclude in sync_list
I started very limited with my sync_list, and now I want to make it a bit more complex. 
I think I understand how it works, but to be sure, are these statement correct?
- The use of sync_list excludes _everything_ by default.
- Sync_list follows an "exclude overrides include" rule, and requires explicit inclusion.
- Sync_list does NOT follow a "include overrides exclude" rule
- Where you place your excludes in the directory MATTERS for includes in sub directories 
- Order exclusions before inclusions, of course (As according to USAGE.md)

### As an example of "exclude overrides include" rule:
```
# Excludes
!/*.pdf # root directory

# Includes
/Documents/Project/*.pdf # sub-sub-directory
```
This would not work because, we have already excluded ALL pdfs in root directory + all its sub-directories.
### Solution to this example would be:
```
# Includes
/Documents/Project/*.pdf
```
Remove the `!/*.pdf` and focus only on `/Documents/Project/` because everything else is excluded by default.

### As an example of "Where you place your excludes in the directory MATTERS":
I only want photos from Camera Roll on my local computer  
```
# Excludes
!Pictures/

# Includes
Pictures/Camera Roll/*
```
This would **not** work, because we already excluded the `Picture` directory and now cannot access `Pictures/Camera Roll/` directory.
```
# Excludes
!Pictures/*

# Includes
Pictures/Camera Roll/*
``` 
This would work, because we did not exclude `Picture` directory, but all sub-directories.
or we can be more specific, and choose which sub-directories in `Picture` we want to exclude, like this:
```
# Excludes
!Pictures/Saved Pictures
!Pictures/Screenshot Pictures

# Includes
Pictures/Camera Roll/*
``` 
### Conclusion 
I have read the USAGE.md and example of sync_list, but it did not make these points/questions clear to me, but maybe I am making this more complicated than it really is? 

Hopefully someone has the time to answer my question(s) :)  

I have a ton of files and folders in my onedrive, so it is not easy for me test stuff and use the `--dry-run command`, because it takes a lot of time to filter out everything from the client side.
If I can get a solid understanding of how it works, then maybe I can do everything in one go and not waste anymore time trying to get it right.  
I am running v2.4.23 btw.

**Comment by abraunegg** [comment, marked as answer]
@Flakvard 
> The use of sync_list excludes everything by default.

Correct

> Sync_list follows an "exclude overrides include" rule, and requires explicit inclusion.

Correct

> Sync_list does NOT follow a "include overrides exclude" rule

Incorrect. Anything that is specifically *included* is included. This is why ordering is critical.

> Where you place your excludes in the directory MATTERS for includes in sub directories

Correct - but so is rule composition. For example - using your example:
```
# Excludes
!/*.pdf # root directory

# Includes
/Documents/Project/*.pdf # sub-sub-directory
```

Your exclude will *only* exclude PDF files in the root directory and *include* all PDF files in the `/Documents/Project/` directory. So your assumption of 'This would not work because, we have already excluded ALL pdfs in root directory + all its sub-directories.' is *incorrect.*

Additionally, the use of `/` at the start of the line is critically important to determine how a rule is interpreted.

Example:
```
# Excludes
!Pictures/

# Includes
Pictures/Camera Roll/*
```
is NOT the same as:
```
# Excludes
!/Pictures/

# Includes
/Pictures/Camera Roll/*
```
In the first block, you are excluding 'Pictures' *anywhere* but including 'Pictures/Camera Roll/' *anywhere*

In the second block you are excluding '/Pictures' in your OneDrive root folder and including '/Pictures/Camera Roll/' from your OneDrive root folder. In this block '/some_folder/Pictures/' would also be excluded, because, by default, 'sync_list' excludes everything unless explicitly included.

**Reply by Flakvard** [reply]
Thank you for such a quick and detailed response @abraunegg .

Okey wow, this is much more powerful and better than I thought. Fantastic!
The _'anywhere'_ rule (not using `/`) looks very similar to `**` globbing patern for filtering multi-level.
 
Does that mean, as an example:
```
# Excludes
!*.pdf # excludes all pdfs anywhere

# Includes
/Documents/Project/*.pdf # include all pdfs in this directory
```
1. This would **not** work, because we _excluded_ all `*.pdf` files _anywhere_ and that overrides our _include_ in `/Documents/Project/` ?

OR! _as I'm guessing is probably true_ 

2. This **would** work, because we specifically _included_ all `*.pdf` files in `/Documents/Project/` directory, and because of the ordering of _excludes_ and _includes_. 
Also, all other `pdfs` _anywere_ are _excluded_ unless specifically _included_ in `sync_list`.

**Reply by abraunegg** [reply]
Correct - it *should* be included as you have explicitly specified it to be included.

**Reply by abraunegg** [reply]
> Also, all other pdfs anywere are excluded unless specifically included in sync_list.

They are already excluded by *default* not because of the `!*.pdf` rule ... this rule is sort of *redundant* really

**Comment by Flakvard** [comment]
> They are already excluded by default not because of the !*.pdf rule ... this rule is sort of redundant really

aah! yes, of course.
This was very insightful. Again, thank you for taking the time to explain this to me.

I believe that the sync_list is the most important thing for newcomers. It is easy to setup, if you know what you are doing.
That being said, both me and my college started at the same time using this application, we both read the sync_list example and both struggled with really understanding the fundamentals with sync_list. Because we both have so many files on onedrive, using dryrun and testing was painful, and even more painful(and stressful) was cleaning up our mistakes.
Unfortunately he gave up and started using the GUI that Insync made, but maybe I can get him onboard again with this new information :)
 
If you don't mind I would like to contribute to the USAGE.md text. Only the sync_list example and the text around it. To make it even more clear, how it works. Essentially avoid the confusion me and my collage had, when we first read it. 

I have no further questions :)
